pkg/deployments/scaler.go

```diff
@@ -46,69 +70,53 @@ func (s *scaler) UpdateState(replicas, min, max int32) {
 // SetDesiredScale sets the desired scale of the scaler and scales
 // if needed.
 func (s *scaler) SetDesiredScale(n int32) {
-	log.Printf("SetDesiredScale(%v)", n)
-	s.compareScales(-1, s.applyMinMax(n))
-}
-
-func (s *scaler) applyMinMax(n int32) int32 {
-	s.mtx.Lock()
-	min := s.minScale
-	max := s.maxScale
-	s.mtx.Unlock()
-	if n < min {
-		n = min
-	} else if n > max {
-		n = max
-	}
-	return n
-}
-
-func (s *scaler) compareScales(current, desired int32) {
+	log.Printf("SetDesiredScale(%v), current: %v, min: %v, max: %v", n, s.currentScale, s.minScale, s.maxScale)
+	nMinMax := s.applyMinMax(n)
```

Comment:
No need for local var, clearer to do `s.desiredScale = s.applyMinMax(n)`

Reply:
but s.applyMinMax will call `Lock()` so that's why I did this. Is that not needed even when it also calls `Lock()`?

Reply:
I'm pretty sure this would result in a deadlock unless I use the local var, see code example here which simulates what would happen if I were to call s.desiredCale = s.applyMinMax(n) within the Lock()
```

func main() {
	mtx := sync.Mutex{}
	mtx.Lock()
	fmt.Println("Locked")
	mtx.Lock()
	fmt.Println("Double locked")
}
```

Reply:
You can call it before the `Lock()`

Reply:
but I shouldn't modify `s.desiredScale` before calling `Lock()`?

Reply:
after quick chat, I fixed this by removing the lock in applyMinMax and moving the call to applyMinMax into the lock itself.